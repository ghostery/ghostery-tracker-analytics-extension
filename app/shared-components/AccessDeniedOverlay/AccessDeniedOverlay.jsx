import React from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';
import './AccessDeniedOverlay.scss';

const NOT_SIGNED_IN = 1;
const EMAIL_NOT_VERIFIED = 2;
const FREE_TRIAL_EXPIRED = 3;
const ACTIVE = 4;

const AccessDeniedOverlay = (props) => {
  const {
    isFullPage,
    maxWidth,
    userInfo,
    messageCreators,
    emailVerificationSent,
    isPageNotScanned,
  } = props;
  const {
    signedIn, insightsUser, email, emailVerified, freeTrial, freeTrialTriggered,
  } = userInfo;

  let accountState = NOT_SIGNED_IN;
  if (signedIn && (!emailVerified || !freeTrialTriggered)) { accountState = EMAIL_NOT_VERIFIED; }
  if (signedIn && emailVerified && !insightsUser && (!freeTrial && freeTrialTriggered)) {
    accountState = FREE_TRIAL_EXPIRED;
  }
  if (signedIn && emailVerified && (insightsUser || freeTrial)) { accountState = ACTIVE; }
  // TODO implement handling of this state if we ever pass data from the backend
  // indicating if the user has subscribed in the past
  // if (signedIn && !insightsUser) { accountState = NOT_SUBSCRIBED; }

  if (accountState === ACTIVE) {
    return null;
  }

  const logoUrlRoot = 'dist/images/shared';
  let logoUrl;
  let logoAlt;
  let headerCopy;
  let textCopy;
  let buttonCopy;
  let buttonFn;
  let additionalLinks = false;
  switch (accountState) {
    case NOT_SIGNED_IN:
      logoUrl = `${logoUrlRoot}/sign-in.svg`;
      logoAlt = 'Sign In';
      headerCopy = '';
      textCopy = 'Sign in or create an account to continue.';
      buttonCopy = '';
      buttonFn = () => {};
      break;
    case EMAIL_NOT_VERIFIED:
      logoUrl = `${logoUrlRoot}/email-not-verified.svg`;
      logoAlt = 'Verify Email';
      headerCopy = 'Please verify your account to proceed.';
      textCopy = `Click the verification link that was sent to ${email} and refresh Insights.`;
      buttonCopy = emailVerificationSent ? 'Sent' : 'Resend Link';
      buttonFn = emailVerificationSent ? () => {} : messageCreators.sendEmailVerification;
      additionalLinks = true;
      break;
    case FREE_TRIAL_EXPIRED:
      logoUrl = `${logoUrlRoot}/ghosty-insights.svg`;
      logoAlt = 'Ghosty Insights';
      headerCopy = 'Your free trial period has ended.';
      textCopy = 'Consider becoming a Contributor to gain access to Insights.';
      buttonCopy = 'Donate now';
      buttonFn = messageCreators.subscribe;
      break;
    default:
      logoUrl = `${logoUrlRoot}/ghosty-insights.svg`;
      logoAlt = 'Ghosty Insights';
      headerCopy = '';
      textCopy = '';
      buttonCopy = '';
      buttonFn = () => {};
  }

  const ui = {
    image: {
      src: chrome.extension.getURL(logoUrl),
      alt: logoAlt,
    },
    header: { copy: headerCopy },
    text: { copy: textCopy },
    button: {
      copy: buttonCopy,
      fn: buttonFn,
    },
  };

  const containerClassNames = ClassNames(
    'AccessDeniedOverlay__container',
    'd-flex',
    'align-items-center',
    'flex-column',
    { altStyling: !signedIn },
  );
  const imageClassNames = ClassNames('AccessDeniedOverlay__image', {
    large: isFullPage,
    emailNotVerified: accountState === EMAIL_NOT_VERIFIED,
  });
  const headerClassNames = ClassNames('AccessDeniedOverlay__header', {
    large: isFullPage,
    reduceMargin: additionalLinks,
  });
  const textClassNames = ClassNames('AccessDeniedOverlay__text', {
    large: isFullPage,
  });
  const buttonClassNames = ClassNames('AccessDeniedOverlay__button', {
    emailSent: signedIn && !emailVerified && emailVerificationSent,
    large: isFullPage,
    reduceMargin: additionalLinks,
  });
  const linkClassNames = ClassNames('AccessDeniedOverlay__link', {
    large: isFullPage,
  });

  return (
    <React.Fragment>
      {!isPageNotScanned && (
        <div className="AccessDeniedOverlay d-flex justify-content-center">
          <div
            className={containerClassNames}
            style={{ maxWidth }}
          >
            <div className="flex-grow-1" />
            <img
              src={ui.image.src}
              className={imageClassNames}
              alt={ui.image.alt}
            />
            {ui.header.copy && (
              <div className={headerClassNames}>
                {ui.header.copy}
              </div>
            )}
            <div className={textClassNames}>
              {ui.text.copy}
            </div>
            {signedIn && (
              <div
                className={buttonClassNames}
                onClick={() => {
                  ui.button.fn();
                  if (accountState === FREE_TRIAL_EXPIRED) {
                    messageCreators.sendMetrics({ type: 'subscribe', insightsView: isFullPage ? '1' : '0' });
                  }
                }}
              >
                {ui.button.copy}
              </div>
            )}
            {additionalLinks && (
              <React.Fragment>
                <div className={linkClassNames} onClick={messageCreators.logout}>
                  Sign Out
                </div>
                <a
                  className={linkClassNames}
                  href="mailto: support-insights@ghostery.com"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  support-insights@ghostery.com
                </a>
              </React.Fragment>
            )}
            <div className="flex-grow-1" />
            {isFullPage && (<div className="flex-grow-1" />)}
          </div>
        </div>
      )}
    </React.Fragment>
  );
};

AccessDeniedOverlay.propTypes = {
  isFullPage: PropTypes.bool,
  maxWidth: PropTypes.string.isRequired,
  userInfo: PropTypes.shape({
    signedIn: PropTypes.bool.isRequired,
    insightsUser: PropTypes.bool.isRequired,
    email: PropTypes.string.isRequired,
    emailVerified: PropTypes.bool.isRequired,
    freeTrial: PropTypes.bool.isRequired,
    freeTrialTriggered: PropTypes.bool.isRequired,
  }).isRequired,
  messageCreators: PropTypes.shape({
    login: PropTypes.func.isRequired,
    logout: PropTypes.func.isRequired,
    subscribe: PropTypes.func.isRequired,
    sendEmailVerification: PropTypes.func.isRequired,
    sendMetrics: PropTypes.func.isRequired,
  }).isRequired,
  emailVerificationSent: PropTypes.bool.isRequired,
  isPageNotScanned: PropTypes.bool.isRequired,
};

AccessDeniedOverlay.defaultProps = {
  isFullPage: false,
};

export default AccessDeniedOverlay;
