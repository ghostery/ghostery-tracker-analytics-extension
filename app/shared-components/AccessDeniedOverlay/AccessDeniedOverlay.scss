@import '../../utils/scss/color-variables.scss';
@import '../../utils/scss/z-index-variables.scss';

.AccessDeniedOverlay {
  position: absolute;
  z-index: $z-index-access-denied-overlay;
  background-color: $opaque-background-darker;
  width: 100%;
  height: 100%;
  text-align: center;

  // Make the logo bright in the Panel

  + .Panel__container .Header__logo > img {
    z-index: $z-index-access-denied-brand-icon;
    pointer-events: none;
  }

  // Make the logo bright in the Insights App

  + .App__container .Header__logo > img {
    z-index: $z-index-access-denied-brand-icon;
    pointer-events: none;
  }
}

.AccessDeniedOverlay__container {
  height: 100%;

  &.altStyling {
    position: relative;
    top: 22px;
    right: 145px;
  }
}

.AccessDeniedOverlay__image {
  margin-bottom: 16px;
  height: 46px;

  &.large {
    margin-top: 60px;
    margin-bottom: 30px;
    height: 56px;
  }

  &.emailNotVerified {
    margin-left: 10px;
    height: 38px;
  }
}

.AccessDeniedOverlay__header {
  margin-bottom: 24px;
  font-weight: 700;
  font-size: 14px;

  &.large {
    margin-bottom: 50px;
    font-size: 18px;
  }

  &.reduceMargin {
    margin-bottom: 15px;
  }

  &.reduceMargin.large {
    margin-bottom: 25px;
  }
}

.AccessDeniedOverlay__text {
  font-weight: 600;
  font-size: 14px;
  line-height: 28px;
  color: $white;

  &.large {
    font-size: 18px;
  }
}

.AccessDeniedOverlay__button {
  margin-top: 50px;
  background: $background-button-gradient;
  border-radius: 4px;
  padding: 8px 28px;
  font-size: 16px;
  line-height: 21px;
  color: $white;
  cursor: pointer;

  &.emailSent {
    background: $dark-grey;
    cursor: default;
  }

  &.large {
    margin-top: 64px;
  }

  &.reduceMargin {
    margin-top: 18px;
    margin-bottom: 10px;
  }

  &.reduceMargin.large {
    margin-top: 39px;
    margin-bottom: 34px;
  }
}

.AccessDeniedOverlay__link {
  margin-top: 10px;
  text-decoration: underline;
  color: $white;
  cursor: pointer;

  &:visited {
    color: white;
  }

  &.large {
    margin: 0;
  }

  &.large:last-of-type {
    margin-top: 15px;
  }
}
