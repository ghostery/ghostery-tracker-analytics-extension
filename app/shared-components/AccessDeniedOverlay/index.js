import { connect } from 'react-redux';
import {
  login,
  logout,
  subscribe,
  sendMetrics,
  sendEmailVerification,
} from '../../store/messages/messageCreators';
import AccessDeniedOverlay from './AccessDeniedOverlay.jsx';

const mapStateToProps = state => ({
  userInfo: state.SettingsReducer.userInfo,
  emailVerificationSent: state.UserInterfaceReducer.emailVerificationSent,
  messageCreators: { login, logout, subscribe, sendMetrics, sendEmailVerification },
  isPageNotScanned: state.PageInfoReducer.isPageNotScanned,
});

export default connect(mapStateToProps)(AccessDeniedOverlay);
