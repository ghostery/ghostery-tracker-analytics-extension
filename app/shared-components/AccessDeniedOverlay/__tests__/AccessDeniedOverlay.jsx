import React from 'react';
import { shallow } from 'enzyme';
import chrome from 'sinon-chrome';

import AccessDeniedOverlay from '../AccessDeniedOverlay';

describe('app/shared-componenets/AccessDeniedOverlay/', () => {
  describe('Snapshot tests with shallow render', () => {
    test('AccessDeniedOverlay component should render with state: Not Signed In', () => {
      global.chrome = chrome;
      const AccessDeniedOverlayProps = {
        maxWidth: '500',
        userInfo: {
          signedIn: true,
          insightsUser: false,
          email: 'test@gmail.com',
          emailVerified: false,
          freeTrial: false,
          freeTrialTriggered: false,
        },
        messageCreators: {
          login: () => {},
          logout: () => {},
          subscribe: () => {},
          sendEmailVerification: () => {},
          sendMetrics: () => {},
        },
        emailVerificationSent: true,
        isPageNotScanned: false,
      };

      const component = shallow(<AccessDeniedOverlay {...AccessDeniedOverlayProps} />);

      expect(component).toMatchSnapshot();
    });
  });

  test('AccessDeniedOverlay component should render with state: FREE_TRIAL_EXPIRED', () => {
    global.chrome = chrome;
    const AccessDeniedOverlayProps = {
      maxWidth: '500',
      userInfo: {
        signedIn: true,
        insightsUser: false,
        email: 'test@gmail.com',
        emailVerified: true,
        freeTrial: false,
        freeTrialTriggered: true,
      },
      messageCreators: {
        login: () => {},
        logout: () => {},
        subscribe: () => {},
        sendEmailVerification: () => {},
        sendMetrics: () => {},
      },
      emailVerificationSent: true,
      isPageNotScanned: false,
    };

    const component = shallow(<AccessDeniedOverlay {...AccessDeniedOverlayProps} />);

    expect(component.find('div.AccessDeniedOverlay__button').length).toBe(1);
  });
  describe('Click functionality testing', () => {
    test('Click Button is present ', () => {
      global.chrome = chrome;

      const AccessDeniedOverlayProps = {
        maxWidth: '500',
        userInfo: {
          signedIn: true,
          insightsUser: false,
          email: 'test@gmail.com',
          emailVerified: true,
          freeTrial: false,
          freeTrialTriggered: true,
        },
        messageCreators: {
          login: () => {},
          logout: () => {},
          subscribe: () => {},
          sendEmailVerification: () => {},
          sendMetrics: () => {},
        },
        emailVerificationSent: true,
        isPageNotScanned: false,
      };

      const component = shallow(<AccessDeniedOverlay {...AccessDeniedOverlayProps} />);
      expect(component.find('div.AccessDeniedOverlay__button')).toHaveLength(1);
    });
  });
});
