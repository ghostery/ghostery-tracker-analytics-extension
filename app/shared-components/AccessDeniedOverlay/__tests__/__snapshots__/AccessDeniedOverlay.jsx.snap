// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/shared-componenets/AccessDeniedOverlay/ Snapshot tests with shallow render AccessDeniedOverlay component should render with state: Not Signed In 1`] = `
<Fragment>
  <div
    className="AccessDeniedOverlay d-flex justify-content-center"
  >
    <div
      className="AccessDeniedOverlay__container d-flex align-items-center flex-column"
      style={
        Object {
          "maxWidth": "500",
        }
      }
    >
      <div
        className="flex-grow-1"
      />
      <img
        alt="Verify Email"
        className="AccessDeniedOverlay__image emailNotVerified"
      />
      <div
        className="AccessDeniedOverlay__header reduceMargin"
      >
        Please verify your account to proceed.
      </div>
      <div
        className="AccessDeniedOverlay__text"
      >
        Click the verification link that was sent to test@gmail.com and refresh Insights.
      </div>
      <div
        className="AccessDeniedOverlay__button emailSent reduceMargin"
        onClick={[Function]}
      >
        Sent
      </div>
      <div
        className="AccessDeniedOverlay__link"
        onClick={[Function]}
      >
        Sign Out
      </div>
      <a
        className="AccessDeniedOverlay__link"
        href="mailto: support-insights@ghostery.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        support-insights@ghostery.com
      </a>
      <div
        className="flex-grow-1"
      />
    </div>
  </div>
</Fragment>
`;
